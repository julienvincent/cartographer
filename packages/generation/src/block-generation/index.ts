export * from './block-space';
export * from './staircase';
