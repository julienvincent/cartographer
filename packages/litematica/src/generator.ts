import * as litematica_bit_array from './litematica-bit-array';
import * as _ from 'lodash';

type PaletteBlockProperty = {
  type: 'string';
  value: string;
};
type PaletteBlockProperties = {
  type: 'compound';
  value: Record<string, PaletteBlockProperty>;
};
type PaletteBlock = {
  Name: {
    type: 'string';
    value: string;
  };
  Properties?: PaletteBlockProperties;
};

export type Block = {
  id: string;
  properties?: Record<string, string>;
};

export type Coords = {
  x: number;
  y: number;
  z: number;
};

export type BlockPosition = Block & Coords;

const createPaletteBlock = (block: Block): PaletteBlock => {
  const palette_block: PaletteBlock = {
    Name: {
      type: 'string',
      value: block.id
    }
  };

  if (Object.keys(block.properties || {}).length > 0) {
    palette_block.Properties = {
      type: 'compound',
      value: _.mapValues(block.properties, (value) => {
        return {
          type: 'string',
          value: value
        } as PaletteBlockProperty;
      })
    };
  }

  return palette_block;
};

const comparePaletteBlocks = (a: PaletteBlock, b: PaletteBlock) => {
  if (a.Name.value !== b.Name.value) {
    return false;
  }

  if (a.Properties) {
    const keys = Object.keys(a.Properties.value);
    if (!b.Properties || Object.keys(b.Properties.value).length !== keys.length) {
      return false;
    }
    return keys.reduce((matching, key) => {
      if (!matching) {
        return false;
      }
      return a.Properties!.value[key].value === b.Properties!.value[key]?.value;
    }, true);
  } else {
    if (b.Properties) {
      return false;
    }
  }

  return true;
};

const createIndexKeyFromCoords = (block: Coords) => {
  return `${block.x}:${block.y}:${block.z}`;
};

const max = <T>(coll: T[], accessor: (item: T) => number) => {
  return coll.reduce((acc, item) => {
    const val = accessor(item);
    if (val > acc) {
      return val;
    }
    return acc;
  }, accessor(coll[0]));
};

type Params = {
  blocks: BlockPosition[];
  name?: string;
  description?: string;
  author?: string;
};
export const generateSchematicNBT = (params: Params) => {
  const length = max(params.blocks, (position) => position.z) + 1;
  const width = max(params.blocks, (position) => position.x) + 1;
  const height = max(params.blocks, (position) => position.y) + 1;

  const volume = length * width * height;
  const AIR = createPaletteBlock({
    id: 'minecraft:air'
  });

  const palette = params.blocks.reduce(
    (palette, block) => {
      const palette_block = createPaletteBlock(block);
      const exists = palette.find((existing) => {
        return comparePaletteBlocks(palette_block, existing);
      });
      if (!exists) {
        palette.push(palette_block);
      }
      return palette;
    },
    [AIR]
  );

  const index = new Map<string, Block>(
    params.blocks.map((block) => {
      return [createIndexKeyFromCoords(block), block];
    })
  );

  const bit_array = _.range(width).reduce((bit_array, x) => {
    return _.range(length).reduce((bit_array, z) => {
      return _.range(height).reduce((bit_array, y) => {
        const block = index.get(createIndexKeyFromCoords({ x, y, z }));

        const palette_index = palette.findIndex((item) => {
          if (block) {
            return comparePaletteBlocks(item, createPaletteBlock(block));
          }
          return comparePaletteBlocks(item, AIR);
        });

        const block_coords = (y * length + z) * width + x;
        return litematica_bit_array.set(bit_array, block_coords, palette_index);
      }, bit_array);
    }, bit_array);
  }, litematica_bit_array.createBitArray(volume, palette.length));

  return {
    name: params.name ?? '',
    value: {
      MinecraftDataVersion: {
        type: 'int',
        value: 2584
      },
      Version: {
        type: 'int',
        value: 5
      },
      Metadata: {
        type: 'compound',
        value: {
          TimeCreated: {
            type: 'long',
            value: [375, 2023476248] // TODO: Use correct time
          },
          TimeModified: {
            type: 'long',
            value: [375, 2023476248] // TODO: Use correct time
          },
          EnclosingSize: {
            type: 'compound',
            value: {
              x: {
                type: 'int',
                value: width
              },
              y: {
                type: 'int',
                value: height
              },
              z: {
                type: 'int',
                value: length
              }
            }
          },
          Description: {
            type: 'string',
            value: params.description ?? ''
          },
          RegionCount: {
            type: 'int',
            value: 1
          },
          TotalBlocks: {
            type: 'int',
            value: palette.length - 1
          },
          Author: {
            type: 'string',
            value: params.author ?? ''
          },
          TotalVolume: {
            type: 'int',
            value: volume
          },
          Name: {
            type: 'string',
            value: params.name ?? ''
          }
        }
      },
      Regions: {
        type: 'compound',
        value: {
          map: {
            type: 'compound',
            value: {
              BlockStates: {
                type: 'longArray',
                value: bit_array.array
              },
              PendingBlockTicks: {
                type: 'list',
                value: {
                  type: 'end',
                  value: []
                }
              },
              Position: {
                type: 'compound',
                value: {
                  x: {
                    type: 'int',
                    value: 1
                  },
                  y: {
                    type: 'int',
                    value: 0
                  },
                  z: {
                    type: 'int',
                    value: 1
                  }
                }
              },
              BlockStatePalette: {
                type: 'list',
                value: {
                  type: 'compound',
                  value: palette
                }
              },
              Size: {
                type: 'compound',
                value: {
                  x: {
                    type: 'int',
                    value: width
                  },
                  y: {
                    type: 'int',
                    value: height
                  },
                  z: {
                    type: 'int',
                    value: length
                  }
                }
              },
              PendingFluidTicks: {
                type: 'list',
                value: {
                  type: 'end',
                  value: []
                }
              },
              TileEntities: {
                type: 'list',
                value: {
                  type: 'end',
                  value: []
                }
              },
              Entities: {
                type: 'list',
                value: {
                  type: 'end',
                  value: []
                }
              }
            }
          }
        }
      }
    }
  };
};
