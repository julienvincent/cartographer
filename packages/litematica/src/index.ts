export * as litematica_bit_array from './litematica-bit-array';
export * as long from './long';
export * from './generator';
