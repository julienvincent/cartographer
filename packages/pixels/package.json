{
  "name": "@cartographer/pixels",
  "version": "0.0.1",
  "private": true,
  "main": "dist/index",
  "types": "dist/index",
  "scripts": {
    "build": "tsc -b",
    "test": "jest --verbose"
  },
  "dependencies": {
    "lodash": "^4.17.21",
    "nearest-color": "^0.4.4",
    "tinycolor2": "^1.4.2"
  },
  "devDependencies": {
    "@types/lodash": "^4.14.191",
    "@types/tinycolor2": "^1.4.3"
  }
}
