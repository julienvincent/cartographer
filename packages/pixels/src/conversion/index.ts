export * from './pixel-data';
export * from './mc-blocks';
