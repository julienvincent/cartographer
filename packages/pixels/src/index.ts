export * as transformers from './transformers';
export * as conversion from './conversion';
export * from './definitions';
