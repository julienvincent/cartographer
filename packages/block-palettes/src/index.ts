import { BlockPalette } from '@cartographer/pixels';

import palette_1_20 from './palettes/1.20.json';
import palette_1_19 from './palettes/1.19.json';
import palette_1_18 from './palettes/1.18.json';

export const palettes = {
  '1.20': palette_1_20 as unknown as BlockPalette,
  '1.19': palette_1_19 as unknown as BlockPalette,
  '1.18': palette_1_18 as unknown as BlockPalette
};
