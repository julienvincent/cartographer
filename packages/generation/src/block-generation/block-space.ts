import { generateOptimizedStaircase } from './staircase';
import * as pixels from '@cartographer/pixels';

export type MCBlockSpace = pixels.MCBlockWithOffset[][][];

const defaultNorthBorderBlock: pixels.MCBlockWithOffset = {
  id: 'minecraft:stone',
  offset: 1
};

export const buildBlockSpace = (mcBlockGrid: pixels.BlockGrid): MCBlockSpace => {
  const space: MCBlockSpace = [];
  const width: number = mcBlockGrid[0].length;
  const height: number = mcBlockGrid.length;

  for (let x = 0; x < width; x++) {
    const strip = mcBlockGrid.map((a) => a[x]); // to redo
    const staircase: number[] = generateOptimizedStaircase(strip, (s) => s.offset);

    space[x] = [defaultNorthBorderBlock].concat(strip).map((v, y) => {
      return [
        {
          ...v,
          offset: staircase[y]
        }
      ];
    });
  }

  // apply water fixes
  // apply dirt fixes
  // apply lava fixes?

  return space;
};
