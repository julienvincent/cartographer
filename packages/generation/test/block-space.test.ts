import { describe, it, expect } from 'vitest';

import * as pixels from '@cartographer/pixels';
import * as block_generation from '../src/block-generation';

describe('block-space generation', () => {
  it('should generate a flat block space', () => {
    const block: pixels.MCBlockWithHue = {
      id: 'minecraft:stone',
      hue: 0
    };

    const block_space = block_generation.generateBlockSpace({
      block_grid: [
        [block, block, block],
        [block, block, block]
      ],
      support_block_id: '',
      staircase_alg: block_generation.StaircaseAlgorithm.Continuous
    });

    const largest_offset = block_space.reduce((offset, block) => {
      if (block.y > offset) {
        return block.y;
      }
      return offset;
    }, 0);

    expect(largest_offset).toBe(1);
    expect(block_space).toMatchSnapshot();
  });

  it('should generate a symmetrical staircased block space and include a north block', () => {
    const block: pixels.MCBlockWithHue = {
      id: 'minecraft:stone',
      hue: 1
    };

    const block_space = block_generation.generateBlockSpace({
      block_grid: [
        [block, block, block],
        [block, block, block],
        [block, block, block]
      ],
      support_block_id: 'support',
      staircase_alg: block_generation.StaircaseAlgorithm.Continuous
    });

    const largest_offset = block_space.reduce((offset, block) => {
      if (block.y > offset) {
        return block.y;
      }
      return offset;
    }, 0);

    expect(largest_offset).toBe(4);
    expect(block_space).toMatchSnapshot();
  });

  it('should generate varied staircased block space', () => {
    const block: pixels.MCBlockWithHue = {
      id: 'minecraft:stone',
      hue: 1
    };

    const block_space = block_generation.generateBlockSpace({
      block_grid: [
        [
          block,
          {
            ...block,
            hue: 0
          },
          {
            ...block,
            hue: 0
          }
        ],
        [
          {
            ...block,
            hue: 2
          },
          block,
          {
            ...block,
            hue: 0
          }
        ]
      ],
      support_block_id: 'support',
      staircase_alg: block_generation.StaircaseAlgorithm.Continuous
    });

    const largest_offset = block_space.reduce((offset, block) => {
      if (block.y > offset) {
        return block.y;
      }
      return offset;
    }, 0);

    expect(largest_offset).toBe(4);
    expect(block_space).toMatchSnapshot();
  });

  it('should add support under required blocks', () => {
    const block: pixels.MCBlockWithHue = {
      id: 'minecraft:stone',
      attributes: {
        requires_support: true
      },
      hue: 0
    };

    const block_space = block_generation.generateBlockSpace({
      block_grid: [[block], [block]],
      support_block_id: 'support',
      staircase_alg: block_generation.StaircaseAlgorithm.Continuous
    });

    expect(block_space).toMatchSnapshot();
  });

  it.only('should reset staircase offset to baseline', () => {
    const make = (hue: pixels.BlockHue): pixels.MCBlockWithHue => {
      return {
        id: 'minecraft:stone',
        hue
      };
    };

    const block_space = block_generation.generateBlockSpace({
      block_grid: [
        [make(1)],
        [make(2)],
        [make(1)],
        [make(0)], // Here we should see a reset
        [make(1)],
        [make(2)],
        [make(0)], // Here we should see a reset
        [make(1)]
      ],
      support_block_id: '',
      staircase_alg: block_generation.StaircaseAlgorithm.Baseline
    });

    expect(block_space).toMatchSnapshot();
  });
});
