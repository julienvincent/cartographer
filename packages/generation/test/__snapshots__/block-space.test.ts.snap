// Vitest Snapshot v1

exports[`block-space generation > should reset staircase offset to baseline 1`] = `
[
  {
    "hue": 1,
    "id": "minecraft:stone",
    "x": 0,
    "y": 1,
    "z": 8,
  },
  {
    "hue": 0,
    "id": "minecraft:stone",
    "x": 0,
    "y": 1,
    "z": 7,
  },
  {
    "hue": 2,
    "id": "minecraft:stone",
    "x": 0,
    "y": 0,
    "z": 6,
  },
  {
    "hue": 1,
    "id": "minecraft:stone",
    "x": 0,
    "y": 1,
    "z": 5,
  },
  {
    "hue": 0,
    "id": "minecraft:stone",
    "x": 0,
    "y": 1,
    "z": 4,
  },
  {
    "hue": 1,
    "id": "minecraft:stone",
    "x": 0,
    "y": 0,
    "z": 3,
  },
  {
    "hue": 2,
    "id": "minecraft:stone",
    "x": 0,
    "y": 0,
    "z": 2,
  },
  {
    "hue": 1,
    "id": "minecraft:stone",
    "x": 0,
    "y": 1,
    "z": 1,
  },
  {
    "hue": 0,
    "id": "",
    "x": 0,
    "y": 1,
    "z": 0,
  },
]
`;
